### ISSUE TITLE
Question: is it a good idea to dispatch actions from the newState function?

### ISSUE THREAD

**Issue by antonm76** [ReSwift/ReSwift#474]
Hello!

I have to support a project created with ReSwift and I found that there is number of `StoreSubscriber`'s `newState` functions dispatches new actions from inside themselves. As far as I can see reducers are processed synchronously producing new state while inside the `newState` function we still work with an old copy of the state.

And another question, should I dispatch all my actions and thunks from one specific queue? If yes why wouldn't you serialize it inside the ReSwift lib?

Thanks!

**Comment by mjarvis** [comment]
I would personally avoid dispatching actions in `newState` for performance and logistical reasons, but there is no reason why it should not work.

Reducers are always processed for a given action before subscribers are notified, as ReSwift is entirely synchronous. So a given action dispatched always goes through the `dispatch -> middleware -> reducers -> subscribers` process before another action is dispatched. (ReSwift will error if an action is dispatched in an async manner during this process)

All actions should be dispatched in a synchronous manner. ReSwift itself does not enforce how that should occur, be it by enforcing a single queue in code, vs a wrapper which dispatches, or some sort of locking mechanism. In most circumstances, it is desirable to use only the main thread.

**Comment by antonm76** [comment]
Hey, @mjarvis! That was very quick -)

Okay, I see how I should dispatch my actions.

But I still in doubt about `newState`. Lets imagine the following:
```
func newState(state: State) {
    // logic A
    store.dispatch(action)
    // logic B
}
```
Here `logic B` executes when changes to the state are already applied. Of course if I use `state` parameter to select data and if I do so, then both `logic A` and `logic B` are consistent and both work with the same state.
I also see an option to create an infinite loop if I would dispatch actions from inside the `newState`.

Sorry, it is mostly my dialog with myself -)

**Comment by mjarvis** [comment]
Yes, dispatching an action in `newState` could cause an infinite loop, if it continuously updates selected state and/or you don't have some sort of skipping process in place.

As long as your state is value-typed (struct, etc), then yes, state will behave as you described. If you were to introduce any reference types (classes) in state, you could see different behaviour. This is another edge case why I avoid this kind of dispatching.

Yes, `newState` will be called on the same thread that `dispatch` was called. No thread or queue hopping occurs within ReSwift.

**Comment by antonm76** [comment]
I have made an experiment:
```
func newState(state: state) {
    store.dispatch(action)
}
```
I would expect it to fall into an infinite loop immediately, but it hasn't. Is that because of `subscriptionsAutomaticallySkipRepeats` set to `true` by default?

**Comment by mjarvis** [comment]
Yes.

If the subscribed state is unchanged, with `subscriptionsAutomaticallySkipRepeats: true`, `newState` will not be called.

**Comment by DivineDominion** [comment]
@mjarvis Just curious: How do you write handlers of stuff that dispatch their result, if you avoid dispatching in `newState`? All strictly in middlewares or thunks?

**Comment by dani-mp** [comment]
I find dispatching from `newState` quite odd and never felt like I needed it. If we think in terms of React and Redux—where ReSwift is coming from—the analogy would be to dispatch from `mapStateToProps` or even `render`, which is completely forbidden.

**Comment by dani-mp** [comment]
If you need to put together local state + global state and that's why you dispatch, lift the local state to global, so you can use a reducer for this. You'll avoid creating an additional ReSwifit loop.

**Comment by DivineDominion** [comment]
Hmm I guess it boils down to your description of local vs global state. Like I have a presenter to ask a user for confirmation based of a request that flowed through the state, and then the presenter dispatches a PerformActualEffect action when done -- it's basically a newState→confirm→dispatch setup.

Sometimes it's also for global state: deleting a file → change state to enqueue the files to be deleted → perform deletion → dispatch success/error. (Would now fit a thunk, but I didn't have these back then.)

**Comment by dani-mp** [comment]
With local state I meant state that the view layer holds, in opposition to global state, which is state that the ReSwift store holds.

If we're talking about synchronous operations here, when both pieces of state lives in the store, you should be able to accomplish the same with just once pass, using the original dispatch.

**Comment by mjarvis** [comment]
@DivineDominion Yes, If I need some non-render reaction to an action or state change, I use middleware.
